### PR TITLE
Firecracker: Add support for hot-swapping workspaces

### PR DESCRIPTION
**Pull request by bduffany**
- Split the mounted RW drive into 2 drives: `scratchfs` (persists across container reuse, not hot-swapped for now) and `workspacefs` (hot-swappable)
- Add a `SyncWorkspace` function to allow hot-swapping the workspace device

The `SyncWorkspace` function is not actually called yet because it would wind up doing unnecessary extra work in the hosted bazel case, since hosted bazel actions have no explicit inputs or outputs that they care about. Specifically, we'd unnecessarily nuke the workspace and then re-create an ext4 image with the same contents. Workspaces can be rather large since they can contain an entire git repo, so this would likely slow down hosted bazel too much.

PRs to follow:
* Update the CI runner to write its contents to the scratch disk (at `/root`) instead of the workspace disk.
* Add a config option to specify the `EstimatedFreeDisk` fraction that gets allocated to the scratch disk vs. workspace disk.
* Call `SyncWorkspace` after `DownloadInputs` to get firecracker working for regular bazel actions that want to use runner recycling (specifically, docker-in-firecracker test actions).
* Once that's done, runner-recycling will work for docker-in-firecracker actions, speeding them up significantly since they can reuse the docker image cache from previous tasks and not have to restart the Docker daemon. We'll be able to set `recycle-runner=true` on our Docker workflow actions to make Docker test workflows faster.

Future optimizations to be made:
* Reuse the vmexec client instead of dialing it every time

TODO before submitting:
- [x] Rebuild `initrd.cpio`

---

**Version bump**: Patch 

**Related issues**: https://github.com/buildbuddy-io/buildbuddy-internal/issues/1188

### CODE REVIEW

File: enterprise/server/remote_execution/snaploader/snaploader.go [hunk 1]
```diff
@@ -31,6 +31,10 @@ type LoadSnapshotOptions struct {
 	InitrdImagePath     string
 	ContainerFSPath     string
 
+	// This field is optional -- a snapshot may have a scratch filesystem
+	// attached or it may have on attached at runtime.
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/snaploader/snaploader.go:35]: s/on/one/g

File: enterprise/server/vmexec/vmexec.go [hunk 2]
```diff
@@ -76,6 +76,20 @@ func (x *execServer) Initialize(ctx context.Context, req *vmxpb.InitializeReques
 	return &vmxpb.InitializeResponse{}, nil
 }
 
+func (x *execServer) UnmountWorkspace(ctx context.Context, req *vmxpb.UnmountWorkspaceRequest) (*vmxpb.UnmountWorkspaceResponse, error) {
```
Comment (tylerwilliams) [inline enterprise/server/vmexec/vmexec.go:79]: nit: would be nice to send the vmexec changes in a separate cl

**Review by tylerwilliams — COMMENTED** [review]
still looking at this and gotta run get food but a few quick comments

File: enterprise/server/vmexec/vmexec.go (hunk 2 above)
Comment (bduffany) [reply enterprise/server/vmexec/vmexec.go:79]: Done: https://github.com/buildbuddy-io/buildbuddy/pull/1674

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go [hunk 3]
```diff
@@ -694,6 +698,49 @@ func (c *FirecrackerContainer) LoadSnapshot(ctx context.Context, workspaceDirOve
 	if err != nil {
 		return status.WrapError(err, "Failed to initialize firecracker VM exec client")
 	}
+
+	if workspaceDirOverride != "" {
+		// If the snapshot is being loaded with a different workspaceFS
+		// then handle that now.
+		if err := c.hotSwapWorkspace(ctx, execClient, wsImgRelPath); err != nil {
```
[the remark this answers is not in the document]
Comment (bduffany) [reply enterprise/server/remote_execution/containers/firecracker/firecracker.go:702]: good call -- done

File: enterprise/server/remote_execution/snaploader/snaploader.go (hunk 1 above)
Comment (bduffany) [reply enterprise/server/remote_execution/snaploader/snaploader.go]: Done

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go [hunk 4]
```diff
@@ -700,6 +692,53 @@ func (c *FirecrackerContainer) LoadSnapshot(ctx context.Context, workspaceDirOve
...
+	}
+	c.workspaceFSPath = hostPath
+	return relativePath, nil
+}
+
+// hotSwapWorkspace unmounts the workspace drive from a running firecracker
+// container, updates the workspace block device to an ext4 image pointed to
+// by chrootRelativeImagePath, and re-mounts the drive.
+func (c *FirecrackerContainer) hotSwapWorkspace(ctx context.Context, execClient vmxpb.ExecClient, chrootRelativeImagePath string) error {
+	log.Infof("Hot-swapping guest workspace drive")
```
Comment (vadimberezniker) [inline enterprise/server/remote_execution/containers/firecracker/firecracker.go:732]: Did you mean to leave this in? w/o additional container instance information it's not a very useful log line

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go [hunk 5]
```diff
@@ -700,6 +692,53 @@ func (c *FirecrackerContainer) LoadSnapshot(ctx context.Context, workspaceDirOve
...
+// and returns the chroot-relative path to the created image.
+func (c *FirecrackerContainer) createWorkspaceImage(ctx context.Context, workspacePath string) (string, error) {
+	c.workspaceGeneration++
+	relativePath := fmt.Sprintf("%s.gen_%d.ext4", workspaceFSName, c.workspaceGeneration)
+	hostPath := filepath.Join(c.getChroot(), relativePath)
+	sizeBytes, err := disk.DirSize(workspacePath)
+	if err != nil {
+		return "", err
+	}
+	if err := ext4.DirectoryToImage(ctx, workspacePath, hostPath, sizeBytes+c.constants.DiskSlackSpaceMB*1e6); err != nil {
```
Comment (vadimberezniker) [inline enterprise/server/remote_execution/containers/firecracker/firecracker.go:721]: I'm assuming you will move the configurable "slack space" to the scratch disk with the ci_runner changes to use the scratch disk?

**Review by vadimberezniker — APPROVED** [review]

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go (hunk 5 above)
Comment (bduffany) [reply enterprise/server/remote_execution/containers/firecracker/firecracker.go:721]: Yep, that's the plan :)

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go (hunk 4 above)
Comment (bduffany) [reply enterprise/server/remote_execution/containers/firecracker/firecracker.go:732]: Woops, removed. Also added tracing

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go [hunk 6]
```diff
@@ -700,6 +692,58 @@ func (c *FirecrackerContainer) LoadSnapshot(ctx context.Context, workspaceDirOve
...
+	defer span.End()
+
+	if _, err := execClient.UnmountWorkspace(ctx, &vmxpb.UnmountWorkspaceRequest{}); err != nil {
+		return status.WrapError(err, "Failed to unmount workspace")
+	}
+	if err := c.machine.UpdateGuestDrive(ctx, workspaceDriveID, chrootRelativeImagePath); err != nil {
+		return status.InternalErrorf("error updating workspace drive attached to snapshot: %s", err)
+	}
+	if _, err := execClient.MountWorkspace(ctx, &vmxpb.MountWorkspaceRequest{}); err != nil {
+		return status.WrapError(err, "Failed to remount workspace after update")
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/containers/firecracker/firecracker.go:745]: We are terrible at this, and I plan to audit it everywhere at some point, but:

errors message should begin with a lower case letter (because they can be wrapped by other errors higher up)
log lines can begin with an upper case letter (because they are immediately logged and not wrapped)

**Review by tylerwilliams — APPROVED** [review]
LGTM

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go (hunk 6 above)
Comment (bduffany) [reply enterprise/server/remote_execution/containers/firecracker/firecracker.go]: Makes sense, I'll keep this in mind for the future. Fixed the capitalization in the errors added by this PR.
